### PR TITLE
[Hardware][NV] Add support for ModelOpt static scaling checkpoints.

### PR DESCRIPTION
**Pull request by pavanimajety**
This change adds support for a new quantization class that has the
ability to load ModelOpt checkpoints through HF. It can be invoked via -

`llm = LLM(model=model_path, quantization="modelopt")`

The checkpoint needs a hf_quant_config.json for loading the right
quantiztion format.

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by robertgshaw2-neuralmagic** [comment]
With the utilities that I made for fp8, all of the core functionality of fp8 is now refactored out of `fp8.py`.

So, this PR should not need to touch `fp8.py`. Instead, you can use the factored out utils with`ModelOptConfig` and `ModelOptFp8LinearMethod` that mirror the logic we used for AutoFP8

**Comment by robertgshaw2-neuralmagic** [comment]
Utilities are in and this PR should now be unblocked

**Comment by pavanimajety** [comment]
@robertgshaw2-neuralmagic @simon-mo @youkaichao Please review 🙏

**Comment by robertgshaw2-neuralmagic** [comment]
@dsikka going to review

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+    def get_name(cls) -> str:
+        return "modelopt"
+
+    @classmethod
+    def get_supported_act_dtypes(cls) -> List[torch.dtype]:
+        return [torch.bfloat16, torch.half]
+
+    @classmethod
+    def get_min_capability(cls) -> int:
+        return 80
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/modelopt.py:50]: Since this is only FP8 for the moment and FP8 Marlin isn't being used, we should restrict to GPUs with hardware support
```suggestion
        return 89
```

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    @classmethod
+    def get_config_filenames(cls) -> List[str]:
+        return ["hf_quant_config.json"]
+
+    @classmethod
+    def from_config(cls, config: Dict[str, Any]) -> "ModelOptFp8Config":
+        quant_config = cls.get_from_keys(config, ["quantization"])
+        quant_method = quant_config["quant_algo"]
+        is_checkpoint_fp8_serialized = ("FP8" in quant_method)
+        activation_scheme = "static"
```
Comment (mgoin) [inline vllm/model_executor/layers/quantization/modelopt.py:61]: I believe you should check the config matches here

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+        )
+        layer.register_parameter("weight", weight)
+        set_weight_attrs(
+            weight,
+            {
+                **extra_weight_attrs,
+                "input_dim": 1,
+                "output_dim": 0,
+            },
+        )
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/modelopt.py:139]: Instead of using a raw torch.nn.Parameter, `ModelWeightParameter` should be used: https://github.com/vllm-project/vllm/blob/2188a60c7e0e5a414a87a4f0fd798333b2e0f625/vllm/model_executor/parameter.py#L166
```
weight = ModelWeightParameter(
data=torch.empty(
                output_size_per_partition,
                input_size_per_partition,
                dtype=weight_dtype,
            ),
input_dim=1,
output_dim=0,
weight_loader=weight_loader
)
```

This will eliminate the need for `set_weight_attrs`

See fp8 for an example: https://github.com/vllm-project/vllm/blob/2188a60c7e0e5a414a87a4f0fd798333b2e0f625/vllm/model_executor/layers/quantization/fp8.py#L155

File: vllm/model_executor/layers/quantization/modelopt.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+        # If checkpoint is fp8, load them.
+        # Otherwise, wait until process_weights_after_loading.
+        if self.quant_config.is_checkpoint_fp8_serialized:
+            # WEIGHT SCALE
+            weight_scale = create_per_tensor_scale_param(
```
Comment (dsikka) [inline vllm/model_executor/layers/quantization/modelopt.py:145]: Similarly, these functions will be removed soon. `PerTensorScaleParameter` should be used instead.

Example: 

https://github.com/vllm-project/vllm/blob/2188a60c7e0e5a414a87a4f0fd798333b2e0f625/vllm/model_executor/layers/quantization/fp8.py#L177

File: vllm/model_executor/layers/quantization/modelopt.py (hunk 2 above)
Comment (pavanimajety) [reply vllm/model_executor/layers/quantization/modelopt.py:61]: Config doesn't record "static" or "dynamic" with Model Optimizer config at the moment since it is only relevant for static quantization:
```
{
    "producer": {
        "name": "modelopt",
        "version": "0.17.0"
    },
    "quantization": {
        "quant_algo": "FP8",
        "kv_cache_quant_algo": null
    }
}

```

Removing references to "static" since it doesn't look relevant for this quantization class.

File: tests/models/test_modelopt.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+from transformers import AutoTokenizer
+
+from tests.quantization.utils import is_quant_method_supported
+from vllm import LLM, SamplingParams
+
+os.environ["TOKENIZERS_PARALLELISM"] = "true"
+
+MAX_MODEL_LEN = 1024
+
+MODELS = ["nvidia/Llama-3.1-8B-Instruct-FP8"]
```
Comment (mgoin) [inline tests/models/test_modelopt.py:18]: Just leaving a note that this model doesn't exist yet

**Review by mgoin — APPROVED** [review]
This looks good enough for now to land! Looking forward to a model checkpoint to test. You may want to merge with main to get the CI to pass

**Review by dsikka — APPROVED** [review]
LGTM
